lib/core/builtin_headers/intrinsic_impl.cpp

```diff
@@ -0,0 +1,71 @@
+#include "core/builtin_headers/intrinsic_impl.h"
+#include "core/builtin_headers/okl_intrinsic_cuda.h"
+#include "core/builtin_headers/okl_intrinsic_dpcpp.h"
+#include "core/builtin_headers/okl_intrinsic_hip.h"
+#include "core/builtin_headers/okl_intrinsic_host.h"
+
+#include "core/transpiler_session/transpiler_session.h"
+#include "core/vfs/overlay_fs.h"
+#include <clang/Frontend/CompilerInstance.h>
+
+namespace oklt {
+
+struct IntrinsicInfo {
+    std::string source;
+    std::vector<std::string> includes;
+};
+
+IntrinsicInfo getIntrinsicInfo(TargetBackend backend) {
+    switch (backend) {
+        case TargetBackend::CUDA:
+            return IntrinsicInfo {
+                std::string(INTRINSIC_CUDA),
+                {"cuda_pipeline_primitives.h"}
+            };
+        case TargetBackend::DPCPP:
+            return IntrinsicInfo {
+                std::string(INTRINSIC_DPCPP),
+                {}
+            };
+        case TargetBackend::HIP:
+            return IntrinsicInfo {
+                std::string(INTRINSIC_HIP),
+                {}
+            };
+        case TargetBackend::OPENMP:
+        case TargetBackend::SERIAL:
+        case TargetBackend::_LAUNCHER:
+            return IntrinsicInfo {
+                std::string(INTRINSIC_HOST),
+                {
+                 "stddef.h",
+                 "cmath"
+                }
+            };
+        default:
+            return {{}, {}};
+    }
+}
+
+void addInstrinsicStub(TranspilerSession &session,
```

Comment:
Does this function work in case of multiple calls?
If not, can it be made to first extract the OverlayFS, and add the file to it?

Reply:
OverlayFileSystem is by design override the files to the most top. Do you think the manual iterating and checking gives more performance instead of just overriding?

Reply:
Did you mean something like this:

```
void addInstrinsicStub(TranspilerSession &session,
                       clang::CompilerInstance &compiler)
{
    auto& headers = session.getStagedHeaders();
    headers.emplace(INTRINSIC_INCLUDE_FILENAME,std::string());
    auto& fm = compiler.getFileManager();
    auto vfs = fm.getVirtualFileSystemPtr();
    if(vfs) {
        std::map<std::string, std::string> not_exists;
        for(const auto &elem: headers) {
            if(!vfs->exists(elem.first)) {
                not_exists.insert(elem);
            }
        }
        auto overlayFs = makeOverlayFs(vfs, not_exists);
        fm.setVirtualFileSystem(overlayFs);
    } else {
        auto overlayFs = makeOverlayFs(vfs, headers);
        fm.setVirtualFileSystem(overlayFs);
    }
}
```

Reply:
> Did you mean something like this:
> 
> ```
> void addInstrinsicStub(TranspilerSession &session,
>                        clang::CompilerInstance &compiler)
> {
>     auto& headers = session.getStagedHeaders();
>     headers.emplace(INTRINSIC_INCLUDE_FILENAME,std::string());
>     auto& fm = compiler.getFileManager();
>     auto vfs = fm.getVirtualFileSystemPtr();
>     if(vfs) {
>         std::map<std::string, std::string> not_exists;
>         for(const auto &elem: headers) {
>             if(!vfs->exists(elem.first)) {
>                 not_exists.insert(elem);
>             }
>         }
>         auto overlayFs = makeOverlayFs(vfs, not_exists);
>         fm.setVirtualFileSystem(overlayFs);
>     } else {
>         auto overlayFs = makeOverlayFs(vfs, headers);
>         fm.setVirtualFileSystem(overlayFs);
>     }
> }
> ```

This example is too complicated and replaces VFS. I'm talking about re-using it.
What about using `vfs->addFile` to add one or multiple files into VFS ?